org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/dfs/DfsBlockCaffeineCache.java

```diff
@@ -0,0 +1,244 @@
+package org.eclipse.jgit.internal.storage.dfs;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+public final class DfsBlockCaffeineCache extends DfsBlockCache {
+
+    /** Pack files smaller than this size can be copied through the cache. */
+    private final long maxStreamThroughCache;
+
+    /**
+     * Suggested block size to read from pack files in bytes.
+     * <p>
+     * If a pack file does not have a native block size, this size will be used.
+     * <p>
+     * If a pack file has a native size, a whole multiple of the native size
+     * will be used until it matches this size.
+     * <p>
+     * The value for blockSize must be a power of 2 and no less than 512.
+     */
+    private final int blockSize;
+
+    /** Cache of pack files, indexed by description. */
+    private final Map<DfsPackDescription, DfsPackFile> packFileCache;
+
+    /** Reverse index from DfsPackKey to the DfsPackDescription. */
+    private final Map<DfsPackKey, DfsPackDescription> reversePackDescriptionIndex;
+
+    /** Cache of Dfs blocks and Indices. */
+    private final Cache<DfsPackKeyWithPosition, Ref> dfsBlockAndIndicesCache;
+
+    public static void reconfigure(final DfsBlockCaffeineCacheConfig cacheConfig) {
+        DfsBlockCache.setInstance(new DfsBlockCaffeineCache(cacheConfig));
+    }
+
+    private DfsBlockCaffeineCache(DfsBlockCaffeineCacheConfig cacheConfig) {
+        maxStreamThroughCache = (long) (cacheConfig.getCacheMaximumSize() * cacheConfig.getStreamRatio());
+
+        blockSize = cacheConfig.getBlockSize();
+
+        packFileCache = new ConcurrentHashMap<>(16, 0.75f, 1);
+        reversePackDescriptionIndex = new ConcurrentHashMap<>(16, 0.75f, 1);
+
+        dfsBlockAndIndicesCache = Caffeine.newBuilder()
+                .removalListener(this::cleanUpIndices)
+                .maximumWeight(cacheConfig.getCacheMaximumSize())
+                // the estimated retained bytes is estimated based on the fields each object contains
+                // plus the key/entry reference itself
+                // DfsPackKey: {
+                //   final int hash;                4 bytes
+                //   final AtomicLong cachedSize;   28 bytes
+                // }
+                // final long position;             8 bytes
+                // final int size;                  4 bytes
+                // key, value reference             8 * 2 bytes
+                .weigher((DfsPackKeyWithPosition keyWithPosition, Ref ref) -> ref == null? 60 : 60 + ref.getSize())
+                .recordStats()
+                .build();
```

Comment:
I think we should make some effort to account for or at least document the memory taken up by the other two maps, e.g. this cache will take ~<= cacheConfig.getCacheMaximumSize() + X MB